## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [6]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [7]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [16]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()

df.dtypes

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object

In [47]:
# What three item categories had the most entries?
top3_categories = df["category"].value_counts().head(3)
top3_categories


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [52]:
# For the category with the most entries, which subcategory had the most entries?
top_category = top3_categories.index[0]

filtered_df = df[df['category'] == top_category]

most_subcategory = filtered_df['subcategory'].value_counts().head(3)
most_subcategory


subcategory
bathroom supplies    6424
paper products       5874
kitchen supplies     5746
Name: count, dtype: int64

In [211]:
# Which five clients had the most entries in the data?
client_count = df['client_id'].value_counts()
client_count_top = client_count.head(5)
client_count_top



client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [148]:
# How many total units (the qty column) did the client with the most entries order order?
total = df[df['client_id'] == 33615]
total_units_ordered = total['qty'].sum()
total_units_ordered


64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [150]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['Subtotal'] = df['unit_price'] * df['qty']
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   Subtotal  
0     1096.80     762.71         7.50  105            1  115164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3    1056.18  
4      195.10     108.17        46.43   20            1    3902.00

In [153]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['Shipping Price'] = df['unit_weight'].apply(lambda x: 7 * x if x > 50 else 10 * x)
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   Subtotal  \
0     1096.80     762.71         7.50  105            1  115164.00   
1       24.95      15.09         1.49   21            0     523.95   
2       13.52       7.86         1.68   39            6     527.28   
3       36.42      24.85         1.23   29            3    1056.18   
4      195.10     108.17        46.43   20            1    3902.00   

   Shipping Price  
0            75.0  
1            14.9  
2            16.8  
3            12.3  
4           464.3

In [195]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['Total Price'] = (df['Subtotal'] + df['Shipping Price']) * 1.0925
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   Subtotal  Shipping Price    Total Price  Line Cost  \
0            1  115164.00            75.0  125898.607500   80159.55   
1            0     523.95            14.9     588.693625     331.79   
2            6     527.28            16.8     594.407400     323.34   
3            3    1056.18            12.3    1167.314400     732.95   
4            1    3902.00           464.3    4770.182750    2627.70   

    Line Profit  
0 -1.196640e+07  
1 -1.042916e+04  
2 -1.998631e+04  
3 -2.947421e+04  
4 -7.373412e+04  

[5 rows x 23 columns]

In [201]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['Line Cost'] = (df['unit_cost'] * df['qty']) + df['Shipping Price']
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   Subtotal  Shipping Price    Total Price  Line Cost  \
0            1  115164.00            75.0  125898.607500   80159.55   
1            0     523.95            14.9     588.693625     331.79   
2            6     527.28            16.8     594.407400     323.34   
3            3    1056.18            12.3    1167.314400     732.95   
4            1    3902.00           464.3    4770.182750    2627.70   

    Line Profit  
0  45732.120000  
1    255.525375  
2    269.513400  
3    433.226650  
4   2099.535000  

[5 rows x 23 columns]

In [197]:
# Create a column for the profit of each line using line cost and line price
df['Line Profit'] = df['Total Price'] - df['Line Cost']
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   Subtotal  Shipping Price    Total Price    Line Cost  \
0            1  115164.00            75.0  125898.607500  80166.48750   
1            0     523.95            14.9     588.693625    333.16825   
2            6     527.28            16.8     594.407400    324.89400   
3            3    1056.18            12.3    1167.314400    734.08775   
4            1    3902.00           464.3    4770.182750   2670.64775   

    Line Profit  
0  45732.120000  
1    255.525375  
2    269.513400  
3    433.226650  
4   2099.535000  

[5 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [205]:
# Check your work using the totals above
order_receipts = {  
    2742071: 152811.89,
    2173913: 162388.71,
    6128929: 923441.25
}

def calculate_total(order_id):
    order_total = df[df['order_id'] == order_id]['Total Price'].sum()
    return order_total

results = {}
for order_id, expected_total in order_receipts.items():
    calculated_total = calculate_total(order_id)
    results[order_id] = {   
        'expected total': expected_total,
        'calculated total': calculated_total
    }
results

{2742071: {'expected total': 152811.89,
  'calculated total': 139844.51202499997},
 2173913: {'expected total': 162388.71,
  'calculated total': 119838.35704999999},
 6128929: {'expected total': 923441.25, 'calculated total': 752748.7454750001}}

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [209]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
cleint_quantities = df.groupby('client_id')['qty'].sum()
top_5_clients = cleint_quantities.sort_values(ascending=False)
top_5_clients.head(5)

client_id
78965    4033302
86721    1010426
30143     846069
85931     644423
77354     506382
Name: qty, dtype: int64

In [228]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
top_5_df = df[df['client_id'].isin(top_5_clients)]
summary_df = top_5_df.groupby('client_id').agg(
    total_units_purchased=('qty', 'sum'),
    total_shipping_price=('Shipping Price', 'sum'),
    total_revenue=('Total Price', 'sum'),
    total_profit=('Line Profit', 'sum')
).reset_index()

summary_df

client_id  total_units_purchased  total_shipping_price  total_revenue  \
0      16892                    359                 330.5   4.096596e+04   
1      23548                  11849                1368.6   2.414331e+06   
2      23906                   6669                1237.4   5.415606e+05   
3      65687                  55546                6348.1   1.002632e+07   
4      74106                   2963                1151.0   1.935194e+05   

   total_profit  
0  1.228926e+04  
1  1.114811e+06  
2  1.207676e+05  
3  3.521878e+06  
4  6.912631e+04

In [229]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 
money_columns = ['total_shipping_price', 'total_revenue', 'total_profit']

# Define a function that converts a dollar amount to millions.
def currency_format_millions(amount):
    return amount / 1000000

# Apply the currency_format_millions function to only the money columns. 
for col in money_columns:
    summary_df[col] = summary_df[col].apply(currency_format_millions)

# Rename the columns to reflect the change in the money format. 
summary_df.rename(columns={
    'total_units_purchased': 'Total Units Purchased',
    'total_shipping_price': 'Total Shipping Price (in millions)',
    'total_revenue': 'Total Revenue (in millions)',
    'total_profit': 'Total Profit (in millions)'
}, inplace= True)

summary_df


client_id  Total Units Purchased  Total Shipping Price (in millions)  \
0      16892                    359                            0.000331   
1      23548                  11849                            0.001369   
2      23906                   6669                            0.001237   
3      65687                  55546                            0.006348   
4      74106                   2963                            0.001151   

   Total Revenue (in millions)  Total Profit (in millions)  
0                     0.040966                    0.012289  
1                     2.414331                    1.114811  
2                     0.541561                    0.120768  
3                    10.026316                    3.521878  
4                     0.193519                    0.069126

In [231]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_summary_df = summary_df.sort_values(by='Total Profit (in millions)', ascending=False)
sorted_summary_df

client_id  Total Units Purchased  Total Shipping Price (in millions)  \
3      65687                  55546                            0.006348   
1      23548                  11849                            0.001369   
2      23906                   6669                            0.001237   
4      74106                   2963                            0.001151   
0      16892                    359                            0.000331   

   Total Revenue (in millions)  Total Profit (in millions)  
3                    10.026316                    3.521878  
1                     2.414331                    1.114811  
2                     0.541561                    0.120768  
4                     0.193519                    0.069126  
0                     0.040966                    0.012289

In [ ]:
# Analysis identified the top five clients by total units purchased and calculated
# their total shipping costs, revenue, and profits. The summary revealed that the client
# with the highest profit also had significant expenditures in both shipping and total purchases,
# highlighting key contributors to the company's profitability. 